setuptools/_distutils_shim.py

```diff
@@ -130,7 +141,8 @@ def frame_file_is_setup(frame):
 
 
 def ensure_shim():
-    DISTUTILS_FINDER in sys.meta_path or add_shim()
+    if not (any(s for s in sys.meta_path if type(s).__name__ == 'DistutilsMetaFinder' and getattr(s, '_fromfile', None) == DistutilsMetaFinder._fromfile)):
+        add_shim()
```

Comment:
Oh, this is unfortunate. Because the shim might be invoked independently several times, it has to detect different instances of the same name and origin. This complexity is one argument for keeping _distutils_hack as a top-level module in order to rely on import semantics to resolve exactly one instance of the module/finder per origin.

I'd like to explore an approach that doesn't require touching some many different aspects of the implementation.

Reply:
That was the point though- since the requirements for the shim are certainly a moving target, the only sure way to get the right shim behavior for any given version of setuptools is to let each setuptools load its own locally, then clean up the others when one actually "fires". It's definitely "icky", but given the restrictions of what you can actually do in a `.pth` file, I can't think of another way that will also accommodate evolving shims.